src/engine.rs

```diff
@@ -2230,4 +2285,128 @@ mod tests {
+    }
+
+    #[test]
+    fn test_start_engine_with_recycle() {
+        let dir = tempfile::Builder::new()
+            .prefix("test_start_engine_with_recycle")
+            .tempdir()
+            .unwrap();
+        let path = dir.path().to_str().unwrap();
+        let cfg = Config {
+            dir: path.to_owned(),
+            target_file_size: ReadableSize(1),
+            purge_threshold: ReadableSize(80), // common size of capacity
+            prefill_for_recycle: true,
+            ..Default::default()
+        };
+        let file_system = Arc::new(DeleteMonitoredFileSystem::new());
+        let engine =
+            RaftLogEngine::open_with_file_system(cfg.clone(), file_system.clone()).unwrap();
+        let (start, end) = engine.file_span(LogQueue::Append);
+        // Only one valid file left, the last one => active_file.
+        assert_eq!(start, end);
+        let stale_count = file_system.inner.file_count() - engine.file_count(None);
+        assert!(stale_count > 0);
+        // Append data.
+        let entry_data = vec![b'x'; 512];
+        for rid in 1..=5 {
+            engine.append(rid, 11, 20, Some(&entry_data));
+        }
+        assert_eq!(engine.file_span(LogQueue::Append).0, start);
+        let (start, end) = engine.file_span(LogQueue::Append);
+        drop(engine);
+        // Reopen the engine with a smaller capacity.
+        let cfg_v2 = Config {
+            target_file_size: ReadableSize::kb(2),
+            purge_threshold: ReadableSize::kb(100),
+            ..cfg
+        };
+        let engine =
+            RaftLogEngine::open_with_file_system(cfg_v2.clone(), file_system.clone()).unwrap();
+        assert_eq!(engine.file_span(LogQueue::Append), (start, end));
+        assert!(stale_count > file_system.inner.file_count() - engine.file_count(None));
+        // Stale files have filled the LogQueue::Append, purge_expired_files won't
+        // truely remove files from it.
+        engine.purge_expired_files().unwrap();
+        assert_eq!(engine.file_span(LogQueue::Append), (start, end));
+        for rid in 1..=10 {
+            engine.append(rid, 20, 31, Some(&entry_data));
+        }
+        assert!(engine.file_span(LogQueue::Append).1 > end);
+        let engine = engine.reopen();
+        assert!(stale_count > file_system.inner.file_count() - engine.file_count(None));
+        let stale_count = file_system.inner.file_count() - engine.file_count(None);
+        drop(engine);
+        // Reopen the engine without log recycling.
+        let cfg_v3 = Config {
+            target_file_size: ReadableSize::kb(2),
+            purge_threshold: ReadableSize::kb(100),
+            enable_log_recycle: false,
+            ..cfg_v2
+        };
+        let engine = RaftLogEngine::open_with_file_system(cfg_v3, file_system.clone()).unwrap();
+        // Restart the engine without log recycling, stale logs should be reserved.
```

Comment:
I thought they should be removed? And if the capacity shrinks, redundant files should be removed too.